### PR TITLE
Alternative fix for #10044

### LINKED ISSUE #10044: REPL: chained extension methods can only be called once
## Minimized code

Only in the REPL, it seems like calling a chain of extension methods has a strange side effect that prevents it from happening again

```Scala
scala> IArray(1,2,3).reverse.sorted                                                                                   
val res0: opaques.IArray[Int] = Array(1, 2, 3)

scala> IArray(1,2,3).reverse.sorted                                                                                   
1 |IArray(1,2,3).reverse.sorted
  |^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |value sorted is not a member of scala.opaques.IArray[Int].
  |An extension method was tried, but could not be fully constructed:
  |
  |    genericArrayOps[T](
  |      opaques.arrayOps.extension_reverse[Int](
  |        {
  |          Array.apply(
  |            {
  |              1
  |            }
  |          , 
  |            {
  |              [2,3 : Int]:Int*
  |            }:Int*
  |          ).asInstanceOf[opaques.IArray[Int]]:IArray[Int]
  |        }
  |      )
  |    )

scala> 
```

## Expectation

this is ok when they are called one after the other in a source file

### PR DESCRIPTION
**Pull request by odersky**
This is a better alternative to fix #10044 than #15120. Unfortunately, it fails
dotty.tools.dotc.BootstrappedOnlyCompilationTests.runWithCompiler in the
tasty interpreter test. Here I see:
```
-- Error: tests/run-custom-args/tasty-interpreter/interpreter/jvm/Interpreter.scala:13:6 -------------------------------
13 |  val jvmReflection = new JVMReflection(using q)
   |      ^
   |compiler bug: created invalid generic signature for getter jvmReflection in scala.tasty.interpreter.jvm.Interpreter
   |signature: ()Lscala/tasty/interpreter/jvm/JVMReflection<()Ljava/lang/Object;>;
   |if this is reproducible, please report bug at https://github.com/lampepfl/dotty/issues
Error while emitting Interpreter.scala
```
If I turn on printing in line 2583 in Types.scala, I see the last output as follows:
```
overwrite TermRef(ThisType(TypeRef(ThisType(TypeRef(NoPrefix,module class jvm)),class Interpreter)),method q) / getter q, class dotty.tools.dotc.core.Denotations$UniqueRefDenotation with getter q at 106
```
So it does overwrite, and the overwrite looks OK, but somehow changing a SymDenotation with a UniqueRefDenotation for the same symbol messes up generic signature generation.

Can someone help me digging deeper here?

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/transform/GenericSignatures.scala [hunk 1]
```diff
@@ -189,7 +189,7 @@ object GenericSignatures {
             fullNameInSig(tp.typeSymbol)
             builder.append(';')
           case _ =>
-            boxedSig(tp)
+            boxedSig(tp.widenDealias.widenNullaryMethod)
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/transform/GenericSignatures.scala:192]:
```suggestion
            boxedSig(tp.widenDealias.widenNullaryMethod)
              // `tp` might be a singleton type referring to a getter.
              // Hence the widenNullaryMethod.
```

**Comment by odersky** [comment]
I think I found and fixed the bug with signature generation.

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 2]
```diff
@@ -2570,9 +2571,25 @@ object Types {
...
+          if denot.validFor.firstPhaseId < ctx.phase.id
+            && lastDenot != null
+            && lastDenot.validFor.lastPhaseId > denot.validFor.firstPhaseId
+            && !lastDenot.isInstanceOf[SymDenotation] =>
+            // In this case the new SymDenotation might be valid for all phases, which means
+            // we would not recompute the denotation when travelling to an earlier phase, maybe
+            // in the next run. We fix that problem by creating a UniqueRefDenotation instead.
+            core.println(i"overwrite ${result.toString} / ${result.lastDenotation}, ${result.lastDenotation.getClass} with $denot at ${ctx.phaseId}")
+            result.setDenot(
+              UniqueRefDenotation(
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/Types.scala:2586]: I think the comment in Compiler.scala about termrefs having only SymDenotation after erasure needs to be updated, but I'm not sure how to reword it to explain what we're doing here:
https://github.com/lampepfl/dotty/blob/main/compiler/src/dotty/tools/dotc/Compiler.scala#L27-L33

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 3]
```diff
@@ -2570,9 +2571,25 @@ object Types {
               || adapted.info.eq(denot.info))
             adapted
           else this
-        result.setDenot(denot)
+        val lastDenot = result.lastDenotation
+        denot match
+          case denot: SymDenotation
+          if denot.validFor.firstPhaseId < ctx.phase.id
+            && lastDenot != null
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/Types.scala:2578]: I'm not completely sure that we won't run into issues where lastDenot didn't exist, but we need to get the denotation at an earlier phase in a subsequent run or via time-travelling. How expensive would it be to always create UniqueRefDenotation after erasure?

**Review by smarter — CHANGES_REQUESTED** [review]
Otherwise LGTM, I rebased and applied the suggested comment.

File: compiler/src/dotty/tools/dotc/core/Types.scala (hunk 3 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/Types.scala:2578]: I prefer caution here. I don't have the time to look into a change like this, which looks risky to me.

File: compiler/src/dotty/tools/dotc/core/Types.scala (hunk 2 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/Types.scala:2586]: done
